src/queuing/single_queue_manager.h

```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * Check whether a worker exists capable of processing given request (according to headers)
+     * @param request_ptr request to be tested
+     * @return true if some worker exists, false if the request cannot be accomodated with current workers
+     */
+    bool is_request_assignable(request_ptr request) const
+    {
+        for (auto &worker : workers_) {
+            if (worker->check_headers(request->headers)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+public:
+    explicit single_queue_manager():
+        comparator_(std::make_unique<JobComparator>()), selector_(std::make_unique<IdleWorkerSelector>())
+    {}
+
+    explicit single_queue_manager(std::unique_ptr<JobComparator> comparator):
+        comparator_(std::move(comparator)), selector_(std::make_unique<IdleWorkerSelector>())
+    {}
+
+    single_queue_manager(std::unique_ptr<JobComparator> comparator, std::unique_ptr<IdleWorkerSelector> selector):
+        comparator_(std::move(comparator)), selector_(std::move(selector))
+    {}
+
+    ~single_queue_manager() override = default;
+
+    request_ptr add_worker(worker_ptr worker, request_ptr current_request = nullptr) override
+    {
+        worker_jobs_[worker] = current_request;
+        workers_.push_back(worker);
+
+        if (current_request != nullptr) {
+            return current_request;
+        }
+
+        return assign_request(worker);
+    }
+
+    request_ptr assign_request(worker_ptr worker) override
+    {
+        worker_jobs_[worker] = nullptr;
+        comparator_->drop_caches();
+
+        std::sort(jobs_.begin(), jobs_.end(), [this, worker] (const request_entry &a, const request_entry &b) {
+            return comparator_->compare(a, b, worker);
+        });
+
+        for (auto it = jobs_.cbegin(); it != jobs_.cend(); ++it) {
+            if (!worker->check_headers(it->request->headers)) {
+                continue;
+            }
+
+            worker_jobs_[worker] = it->request;
+            jobs_.erase(it);
+
+            return worker_jobs_[worker];
+        }
+
+        return nullptr;
+    }
+
+    std::shared_ptr<std::vector<request_ptr>> worker_terminated(worker_ptr worker) override
+    {
+        auto result = std::make_shared<std::vector<request_ptr>>();
+        if (worker_jobs_[worker] != nullptr) {
+            result->push_back(worker_jobs_[worker]); // currently running job (returned for possible reasignment)
+        }
+        worker_jobs_.erase(worker);
+        workers_.erase(std::remove(workers_.begin(), workers_.end(), worker), workers_.end());
+
+        // filter jobs and remove those wich are no longer process-able (after worker removal)
+        for (auto &&job : jobs_) {
+            if (!is_request_assignable(job.request)) {
+                // the job cannot be accomodated anymore...
+                result->push_back(job.request);
+                job.request = nullptr; // mark the job for removal
```

Comment:
That's... unusual. In such short code though... I guess it's alright

Reply:
What is unusual? The marking for deletion? If you have a better way, I would be happy to entertain it, but I am quite sure that deleting items from a container in a for-each loop (over that container) is not something you want to do...

Reply:
If I went with the "mark for deletion" route, I'd probably make a separate collection of stuff to remove. As I said, the current implementation works just fine and it's unlikely the method will ever grow enough for it to become unmaintainable.

An alternative idea I have is to use `stable_partition` to push the unsatisfiable request to the edge of the queue and then snip them off.

Reply:
that's what remove_if does ... I guess I have become too biased towards functional JS-like solutions ;)

Reply:
well, you can call mutating the queue items in place many names, but functional isn't one of them :slightly_smiling_face:

Reply:
Yep, I mutated the objects inside, but the collection was intact until filtered by erase-remove, which is functional (that's what I meant).